lib/ServiceNameResolver.h

```diff
@@ -29,29 +29,33 @@ namespace pulsar {
 class ServiceNameResolver {
    public:
     ServiceNameResolver(const std::string& uriString)
-        : serviceUri_(uriString), numAddresses_(serviceUri_.getServiceHosts().size()) {
+        : serviceUri_(std::make_shared<ServiceURI>(uriString)),
+          numAddresses_(serviceUri_->getServiceHosts().size()) {
         assert(numAddresses_ > 0);  // the validation has been done in ServiceURI
     }
 
     ServiceNameResolver(const ServiceNameResolver&) = delete;
     ServiceNameResolver& operator=(const ServiceNameResolver&) = delete;
 
     bool useTls() const noexcept {
-        return serviceUri_.getScheme() == PulsarScheme::PULSAR_SSL ||
-               serviceUri_.getScheme() == PulsarScheme::HTTPS;
+        return serviceUri_->getScheme() == PulsarScheme::PULSAR_SSL ||
+               serviceUri_->getScheme() == PulsarScheme::HTTPS;
     }
 
     bool useHttp() const noexcept {
-        return serviceUri_.getScheme() == PulsarScheme::HTTP ||
-               serviceUri_.getScheme() == PulsarScheme::HTTPS;
+        return serviceUri_->getScheme() == PulsarScheme::HTTP ||
+               serviceUri_->getScheme() == PulsarScheme::HTTPS;
     }
 
     const std::string& resolveHost() {
-        return serviceUri_.getServiceHosts()[(numAddresses_ == 1) ? 0 : (index_++ % numAddresses_)];
+        return serviceUri_->getServiceHosts()[(numAddresses_ == 1) ? 0 : (index_++ % numAddresses_)];
     }
 
+    void updateServiceUrl(const std::string& urlString) { serviceUri_.reset(new ServiceURI(urlString)); }
```

Comment:
There is a problem that if the `urlString` contains multiple service URLs, we have to update `numAddresses_` as well. A possible solution is moving `numAddresses_` into `ServiceURI`.